src/metabase/db/liquibase.clj

```diff
@@ -127,16 +128,20 @@
   [conn-or-data-source :- [:or (ms/InstanceOfClass java.sql.Connection) (ms/InstanceOfClass javax.sql.DataSource)]
    f                   :- fn?]
   ;; Custom migrations use toucan2, so we need to make sure it uses the same connection with liquibase
-  (binding [t2.conn/*current-connectable* conn-or-data-source]
-    (if (instance? java.sql.Connection conn-or-data-source)
-      (f (->> conn-or-data-source liquibase-connection database (liquibase conn-or-data-source)))
-      ;; closing the `LiquibaseConnection`/`Database` closes the parent JDBC `Connection`, so only use it in combination
-      ;; with `with-open` *if* we are opening a new JDBC `Connection` from a JDBC spec. If we're passed in a `Connection`,
-      ;; it's safe to assume the caller is managing its lifecycle.
-      (with-open [conn           (.getConnection ^javax.sql.DataSource conn-or-data-source)
-                  liquibase-conn (liquibase-connection conn)
-                  database       (database liquibase-conn)]
-        (f (liquibase conn database))))))
+  (let [f* (fn [liquibase]
+             ;; trigger liquibase to create databasechangelog tables if needed
+             (.checkLiquibaseTables liquibase false (.getDatabaseChangeLog liquibase) nil nil)
+             (f liquibase))]
+    (binding [t2.conn/*current-connectable* conn-or-data-source]
```

Comment:
to be fair, this function only has many lines; what it does is pretty basic. I'm not even sure if we really need 2 branches here, I think it was added for testing purposes.

on macro, I don't think we should put the logic there, our style of writing macro is that it eventually calls `do-something` do the job. the defmacro basically is wrapper of this function.

So I'd rather keep all the logic in this `do-`function rather than scatter it in both places.

Wdyt?

Reply:
Cool - happy to drop this for now. Still have misgivings on this code, and would like to come back an improve this at a later stage. Here are some thoughts.

> To be fair, this function only has many lines; what it does is pretty basic.

Precisely what it does is so basic, I think the code should be as simple as possible. Creating a thunk over a thunk and not executing from top to bottom is complexity that future code readers could be spared in my opinion.

> I'm not even sure if we really need 2 branches here, I think it was added for testing purposes.

1. Test code quality matters too. If anything the bar for being easy to follow is higher.
2. This code is not in a test namespace.
3. Future production code may make use of this distinction.

> On macro, I don't think we should put the logic there.

That's a fair point, I do really like the "always delegate to a function" convention (when it's possible). On the other hand, being able to insert the the side-effect within regular lexical code is much more straightforward than a thunk decorator, and ultimately yields code that's easier to follow in my opinion. 

The "always delegate" convention exists to serve readability, and should be followed so rigorously that it conflicts IMHO. I'll sleep on it in this case, still calibrating my sensibilities.